### DISCUSSION TITLE
[smartcharging] TxDefaultProfile with same stackLevel

### DISCUSSION THREAD

**Discussion by SiarheiVolkau** [c-jimenez/open-ocpp#141] [General]
SmartChargingManager allows installing profiles with connector_id == 0 for `TxDefaultProfile` purpose, although then it allows to apply two profiles with same `stackLevel` and `Purpose` on a connector which is explicitly prohibited by the OCPP-1.6.

> To avoid conflicts, the existence of multiple Charging Profiles with the same stackLevel and Purposes
in a Charge Point is not allowed. Whenever a Charge Point receives a Charging Profile with a
stackLevel and Purpose that already exists in the Charge Point, the Charge Point SHALL replace the
existing profile.

While OCPP doesn't state how to implement installing TxDefaultProfile for connector 0 it seems more reasonable to install that profile for each connector separately - thus avoiding that conflict situation, doesn't it?

**Comment by c-jimenez** [comment]
Hi,

The case where a TxDefault profile is installed on connector 0 and then a TxDefaultProfile is installed on a specific connector is described in the 3.13.1. Charging profile purposes chapter :

```
If TxDefaultProfile is set to ConnectorId 0, the TxDefaultProfile is applicable to all Connectors.
If ConnectorId is set >0, it only applies to that specific connector.
In the event a TxDefaultProfile for connector 0 is installed, and the Central System sends a new profile with ConnectorId
>0, the TxDefaultProfile SHALL be replaced only for that specific connector.
```

So you can effectively install a profile with same purpose and stack level on connector 0 and 1 simultaneously.
Moreover in the ClearChargingProfile request you can specify 0 as a connector id so I think this is needed to store a profile for connector 0 (and not duplicate on all connectors) since it can be removed independently of the profiles on other connector.

**Reply by SiarheiVolkau** [reply]
Your arguments make sense and I'm agreeing mostly, although:

> In the event a TxDefaultProfile for connector 0 is installed, and the Central System sends a new profile
with ConnectorId >0, the TxDefaultProfile SHALL be replaced only for that specific connector.

the key phrase here is **SHALL be replaced**, but it doesn't.

Imagine the situation while default profile for connector 0 (`DP0`) is overruled by another one for connector > 0  (`DPn`) and then `DPn` will be cleared. Which one shall take in effect then?
According to specification no one - as `DP0` shall be replaced by `DPn` on connector `n`.

**Reply by c-jimenez** [reply]
Yes you're right, that's a good point.
I will try to get some more information from OCA people to have their opinion on this.
I will keep you inform.
